### PR TITLE
Exchangegcfix

### PR DESCRIPTION
**Pull request by sbird**
Fix Issue #261 by checking for garbage, and a potential issue where domain leaves no space for star formation.

### CODE REVIEW

File: libgadget/domain.c [hunk 1]
```diff
@@ -471,10 +471,11 @@ domain_check_memory_bound(const int print_details, int64_t *TopLeafWork, int64_t
         ta_free(list_load);
     }
 
-    if(max_load > PartManager->MaxPart)
+    /*Leave a small number of particles for star formation */
+    if(max_load > PartManager->MaxPart * 0.99)
```
Comment (rainwoodman) [inline libgadget/domain.c]: I think you want to preserve more. Gut feeling is 0.90.

File: libgadget/exchange.c [hunk 2]
```diff
@@ -169,11 +169,19 @@ static int domain_exchange_once(int (*layoutfunc)(int p), ExchangePlan * plan, i
 
     int bad_exh=0;
 
-    /*Check whether the domain exchange will succeed. If not, bail*/
+    /*Check whether the domain exchange will succeed. If not, check for garbage and then check again.*/
     if(PartManager->NumPart + plan->toGetSum.base - plan->toGoSum.base > PartManager->MaxPart){
-        message(1,"Too many particles for exchange: NumPart=%d count_get = %d count_togo=%d MaxPart=%d\n",
+        /* How much garbage have we got? Keep this serial so there is no memory allocation.
```
Comment (rainwoodman) [inline libgadget/exchange.c]: Is the idea the garbages will be always be freed so we actually have more free slots?

Comment (sbird) [reply libgadget/exchange.c]: That's right - we have to do a gc anyway to get rid of the particles that are leaving this processor, and the garbage gets swept up in that. But hang on, I can do this a better way.

File: libgadget/domain.c (hunk 1 above)
Comment (sbird) [reply libgadget/domain.c]: That seems a little high - remember that only the star formation mode that forks an extra particle will increase the particle count. So we are talking the SH03 model, not the QuickLya star formation module. A 10% increase in particle count would mean 20% of the gas forming stars in one time step, which seems a lot. How about 0.95?

**Comment by sbird** [comment]
That's a better fix, I think.

**Comment by sbird** [comment]
Since this is confirmed to fix the problem, shall I merge?

**Pull request merged**